### PR TITLE
Add gong map class

### PR DESCRIPTION
**Pull request by TrestanSimon**
Adapts the GONG synoptic map class ~~and ADAPT map class~~ originally developed for [pfsspy](https://github.com/dstansby/pfsspy/blob/f493e2cdc653e541425f01318ec2f35645ee36d0/pfsspy/map.py#L12).

Closes https://github.com/sunpy/sunpy/issues/7204.

### CODE REVIEW

**Comment by nabobalis** [comment]
@dstansby do you know if we need to add a licence file from pffspy to ensure the licence change?

**Comment by dstansby** [comment]
No, I think my permission is enough.

@TrestanSimon could you update this PR to remove the ADAPT map class? That wasn't written by me, so will need permission of the other author, and it will be much easier to review this pull request anyway if we just keep it to the GONG map class.

**Comment by TrestanSimon** [comment]
It looks like the codecov tests are failing with a ton of indirect changes to the coverage of unrelated files. This was also happening in a previous PR https://github.com/sunpy/sunpy/pull/7219 that I opened, and I was advised to ignore it.

**Comment by nabobalis** [comment]
The codecov checks looks to have passed on the PR, I wouldn't worry.

**Comment by TrestanSimon** [comment]
> The codecov checks looks to have passed on the PR, I wouldn't worry.

Weird... I see that now. GitHub said it was failing earlier, and the [associated page](https://app.codecov.io/gh/sunpy/sunpy/commit/b7794029fe48c1f8ace0b3860d0433552b392da4/indirect-changes) on the codecov website still lists the unusual indirect coverage changes. Thanks again, I will ignore this going forward.

File: sunpy/map/sources/gong.py
Comment (Cadair) [inline sunpy/map/sources/gong.py:42]: I think the way this `__init__` is written isn't the way we want to handle this type of metadata fixing any more. These fixes should be implemented by overloading the properties on the `GenericMap` class rather than editing the `.header` property in the `__init__`.

For example you would "fix" CUNIT1/2 by writing a property like this: https://github.com/sunpy/sunpy/blob/main/sunpy/map/mapbase.py#L1298

**Review by Cadair — DISMISSED** [review]
Thanks a lot for the PR @TrestanSimon 

I think this needs some rewriting to match the way we have adopted doing metadata fixing in core. Let me know if you need me to give you more details on how to write all the properties.

**Comment by wtbarnes** [comment]
Hi @TrestanSimon 👋 Thanks again for your contribution. We'd like to get this into the next release of sunpy (v5.1) which will tentatively be released on 3 November 2023. Do you think you'll have time to address the comments on this PR by then? If not, one of the maintainers is happy to pick this up and finish it off.

**Comment by TrestanSimon** [comment]
I actually have it done and can commit and push in a few minutes. Apologies for the delay.

File: sunpy/map/sources/gong.py
Comment (ayshih) [inline sunpy/map/sources/gong.py:75]: Um, `observer_coordinate` is supposed to return a `SkyCoord`, not a dictionary...

**Review by ayshih — DISMISSED** [review]

File: sunpy/map/sources/gong.py
Comment (TrestanSimon) [reply sunpy/map/sources/gong.py:75]: I am aware. I mistakenly pushed a work-in-progress version of the code. My final copy of the code is on a different machine I have lost access to (hence the lack of any commits here recently). I thought I had transferred a version with this property addressed before losing access, but I did not.

Regardless, a forthcoming commit should have this implemented properly.

File: sunpy/map/sources/gong.py
Comment (wtbarnes) [inline sunpy/map/sources/gong.py:68]:
```suggestion
```
I don't think we need this property anymore as `u.Unit ('Gauss')` works and the `GenericMap.unit` already covers the None case.

File: sunpy/map/sources/gong.py
Comment (wtbarnes) [inline sunpy/map/sources/gong.py:76]: I'm confused as to why we can't just return `get_earth(self.date)` here.

File: sunpy/map/sources/gong.py
Comment (wtbarnes) [inline sunpy/map/sources/gong.py:52]:
```suggestion
        return Time(f"{self.meta.get('date-obs')} {self.meta.get('time-obs')}")
```

File: sunpy/map/sources/tests/test_gong_synoptic_source.py
Comment (wtbarnes) [inline sunpy/map/sources/tests/test_gong_synoptic_source.py:31]: It's a bit worrying to me that the measurement (which here is just the wavelength) is being returned without a unit. In the above test header, the unit (nm) is included in the key comment, but it does not seem to be being parsed correctly. This is out of scope of this PR though so I'm OK merging and fixing this later.

File: sunpy/map/sources/tests/test_gong_synoptic_source.py
Comment (wtbarnes) [inline sunpy/map/sources/tests/test_gong_synoptic_source.py:44]:
```suggestion
```
I think we can remove this since we don't override the base `unit` property anymore.

**Review by wtbarnes — CHANGES_REQUESTED** [review]
Thanks again for your contribution @TrestanSimon and thanks for your patience during the review. There have been a number of pretty big changes in how metadata is handled since this map source was written, hence all of the comments. Just a few suggested changes and a clarification about the observer coordinate.

File: sunpy/map/sources/gong.py
Comment (TrestanSimon) [reply sunpy/map/sources/gong.py:67]: If we just use ``get_earth(self.date)``, do you think we should test that it is consistent with the metadata?

File: sunpy/map/sources/tests/test_gong_synoptic_source.py
Comment (wtbarnes) [reply sunpy/map/sources/tests/test_gong_synoptic_source.py:31]: See #7245. If that is merged, we can rebase this PR and then change the test to check for `676.8 * u.nm`

File: sunpy/map/sources/gong.py
Comment (wtbarnes) [reply sunpy/map/sources/gong.py:59]: I do not think these GONG maps typically carry around the metadata that specifies the observer location, hence the assumption that the observer is at Earth. Someone with more knowledge of these data products should weigh in though.

File: sunpy/map/sources/gong.py
Comment (wtbarnes) [inline sunpy/map/sources/gong.py:48]:
```suggestion
        # Since, this map uses the cylindrical equal-area (CEA) projection,
        # the spacing should be modified to 180/pi times the original value
        # Reference: Section 5.5, Thompson 2006
```
Sorry, missed this before. Copying over the same comment as https://github.com/sunpy/sunpy/blob/7821330e32384e3d1ee66a411738935364e1ff8b/sunpy/map/sources/sdo.py#L179-L181

**Review by wtbarnes — APPROVED** [review]
Looks great to me, just one minor correction on the comment about the scale correction. Thanks again @TrestanSimon!

**Review by dstansby — APPROVED** [review]

**Pull request merged**
